### PR TITLE
chore(docs): v4 Migration Guide

### PR DESCRIPTION
**Pull request by LekoArts**
## Description

Migration Guide for v4 beta.

[See rendered version](https://github.com/gatsbyjs/gatsby/blob/v4-docs-migration-guide/docs/docs/reference/release-notes/migrating-from-v3-to-v4.md)

### Documentation

Only ready before the actual GA release. Needs adaption then to e.g. rewrite `next` to `latest` etc.

## Related Issues

[ch37888]

### CODE REVIEW

File: docs/docs/reference/release-notes/migrating-from-v3-to-v4.md [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+}
+```
+
+### Removal of obsolete flags
+
+The config flags for `QUERY_ON_DEMAND` and `PRESERVE_WEBPACK_CACHE` have been removed since the features are enabled by default. Thus you can't explicitly opt-out of these anymore.
+
+### Node mutation & creation in custom resolvers
+
+TODO
```
Comment (LekoArts) [inline docs/docs/reference/release-notes/migrating-from-v3-to-v4.md:234]: I'd be grateful about some content here from you @vladar - can be rough/draft, I'll polish it.

Will mark you in other spots as well :)

File: docs/docs/reference/release-notes/migrating-from-v3-to-v4.md [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+## Future Breaking Changes
+
+This section explains deprecations that were made for Gatsby 4. These old behaviors will be removed in v5, at which point they will no longer work. For now, you can still use the old behaviors in v4, but we recommend updating to the new signatures to make future updates easier.
+
+### `nodeModel.runQuery`
+
+This method is available in custom resolvers and returns an unbound array of nodes (without any limit and skip applied). While this is cheap with an in-memory store, with the new LMDB database this is quite expensive. You should instead use `nodeModel.findOne` and `nodeModel.findAll` as they support limit and skip directly.
+
+> TODO: Example code before/after
```
Comment (LekoArts) [inline docs/docs/reference/release-notes/migrating-from-v3-to-v4.md:244]: @vladar

File: docs/docs/reference/release-notes/migrating-from-v3-to-v4.md [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+
+### `nodeModel.getAllNodes`
+
+Similar to `nodeModel.runQuery` the `nodeModel.getAllNodes` returns an unbound array of **all** nodes which is also quite expensive to run. We recommend using `nodeModel.findAll` instead as it at least returns an iterable and not an array.
+
+> TODO: Example code before/after
```
Comment (LekoArts) [inline docs/docs/reference/release-notes/migrating-from-v3-to-v4.md:250]: @vladar

File: docs/docs/reference/release-notes/migrating-from-v3-to-v4.md [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+
+### `___NODE` convention
+
+> TODO: Example code before/after
```
Comment (LekoArts) [inline docs/docs/reference/release-notes/migrating-from-v3-to-v4.md:254]: @vladar

File: docs/docs/reference/release-notes/migrating-from-v3-to-v4.md [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
++   "gatsby": "^3.0.0 || ^4.0.0",
+  }
+}
+```
+
+### Don't mutate data outside of expected APIs
+
+This is an anti-pattern in some plugins that will no longer work with LMDB. You shouldn't mutate data after the `createNode` call. In Gatsby 4 the nodes are only stored once (when sourcing them) and any mutations afterwards are lost.
+
+> TODO: Example of current anti-pattern + proposed better solution
```
Comment (LekoArts) [inline docs/docs/reference/release-notes/migrating-from-v3-to-v4.md:288]: @vladar

**Review by vladar — APPROVED** [review]

**Pull request merged**